src/modules/presence_reginfo/notify_body.c

```diff
@@ -0,0 +1,322 @@
+		}
+		j++;
+	}
+
+	if(j == 0) /* no body */
+	{
+		LM_DBG("no body to be built\n");
+		goto error;
+	}
+
+	/* n: number of bodies in total */
+	/* j: number of useful bodies; created XML structures */
+	/* i: loop counter */
+	/* LM_DBG("number of bodies in total [n]=%d, number of useful bodies [j]=%d\n", n, j ); */
+
+	/* create the new NOTIFY body  */
+	doc = xmlNewDoc(BAD_CAST "1.0");
+	if(doc == 0) {
+		LM_ERR("unable to create xml document\n");
+		goto error;
+	}
+
+	root_node = xmlNewNode(NULL, BAD_CAST "reginfo");
+	if(root_node == 0) {
+		goto error;
+	}
+
+	xmlDocSetRootElement(doc, root_node);
+	namespace = xmlNewNs(
+			root_node, BAD_CAST "urn:ietf:params:xml:ns:reginfo", NULL);
+	if(!namespace) {
+		LM_ERR("creating namespace failed\n");
+	}
+	xmlSetNs(root_node, namespace);
+
+	/* The version must be increased for each new document and is a 32bit int.
+	 * As the version is different for each watcher, we can not set here the
+	 * correct value. Thus, we just put here a placeholder which will be
+	 * replaced by the correct value in the aux_body_processing callback.
+	 * Thus we have CPU intensive XML aggregation only once and can use
+	 * quick search&replace in the per-watcher aux_body_processing callback.
+	 * We use 11 chracters as an signed int (although RFC says unsigned int we
+	 * use signed int as presence module stores "version" in DB as
+	 * signed int) has max. 10 characters + 1 character for the sign
+	 */
+	xmlNewProp(root_node, BAD_CAST "version", BAD_CAST "00000000000");
+	xmlNewProp(root_node, BAD_CAST "state", BAD_CAST "full");
+
+	/* loop over all bodies and create the aggregated body */
+	for(i = 0; i < j; i++) {
+		// LM_DBG("[n]=%d, [i]=%d, [j]=%d xml_array[i]=%p\n", n, i, j,
+		// 		xml_array[j]);
+
+		// get the root reginfo element
+		p_root = xmlDocGetRootElement(xml_array[i]);
+		if(p_root == NULL) {
+			LM_ERR("the xml_tree root element is null\n");
+			goto error;
+		}
+
+		// get the children registration elements
+		if(p_root->children) {
+			// loop over registration elements
+			for(node = p_root->children; node; node = next_node) {
+				next_node = node->next;
+				if(node->type != XML_ELEMENT_NODE) {
+					continue;
+				}
+				LM_DBG("node type: Element, name: %s\n", node->name);
+
+				/* we do not copy the node, but unlink it and then add it ot the new node
+				 * this destroys the original document but we do not need it anyway.
+				 */
+				xmlUnlinkNode(node);
+				if(xmlAddChild(root_node, node) == NULL) {
+					xmlFreeNode(node);
+					LM_ERR("while adding child\n");
+					goto error;
+				}
+			} // end of loop over registration elements
+		}
+	} // end of loop over all bodies
+
+	// convert to string & cleanup
+	xml_array = (xmlDocPtr *)pkg_malloc(n * sizeof(xmlDocPtr));
+	if(xml_array == NULL) {
+		LM_ERR("while allocating memory");
+		return NULL;
+	}
+	memset(xml_array, 0, n * sizeof(xmlDocPtr));
+
+	body = (str *)pkg_malloc(sizeof(str));
+	if(body == NULL) {
+		ERR_MEM(PKG_MEM_STR);
```

Comment:
You could also use PKG_MEM_ERROR, to have it everywhere the same
And you probably also want to exit and cleanup the previous allocated memory, right?

Reply:
Well, ERR_MEM does implicit goto to error where the memory is cleaned. The 2nd alloc of xml_array was a typo and has been removed. Now we have a single xml_array that's freed in the error handling. or not?